cmd/plugins/topology-aware/policy/topology-aware-policy.go

```diff
@@ -148,19 +149,82 @@ func (p *policy) Sync(add []cache.Container, del []cache.Container) error {
 		p.AllocateResources(c)
 	}
 
+	p.checkAllocations("  <post-sync>")
+
 	return nil
 }
 
+func (p *policy) checkAllocations(format string, args ...interface{}) {
+	var (
+		prefix  = fmt.Sprintf(format, args...)
+		cpuExcl = 0
+		cpuPart = 0
+		mem     = uint64(0)
+		ctr     = map[string]Grant{}
+		dup     = map[string][]Grant{}
+	)
+
+	getMemorySize := func(g Grant) uint64 {
+		var (
+			limit = g.MemLimit()
+			total = uint64(0)
+		)
+		for _, memType := range []memoryType{memoryDRAM, memoryPMEM, memoryHBM} {
+			total += limit[memType]
+		}
+		return total
+	}
+
+	for _, g := range p.allocations.grants {
+		log.Debug("%s %s (%s)", prefix, g, g.GetContainer().GetID())
+		full := g.ExclusiveCPUs().Size()
+		part := g.CPUPortion()
+		cpuExcl += full
+		cpuPart += part
+
+		mem += getMemorySize(g)
+
+		_, ok := p.cache.LookupContainer(g.GetContainer().GetID())
+		if !ok {
+			log.Error("%s   %s STALE container among allocations, not found in cache", prefix, g)
+		}
+
+		key := g.GetContainer().PrettyName()
+		old, ok := ctr[key]
+		if ok {
+			if len(dup[key]) == 0 {
+				dup[key] = []Grant{old, g}
+			} else {
+				dup[key] = append(dup[key], g)
+			}
+		} else {
+			ctr[key] = g
+		}
+	}
+
+	for key, grants := range dup {
+		log.Error("%s DUPLICATE allocation entries for container %s", prefix, key)
+		for _, g := range grants {
+			log.Error("%s   %s (%s)", prefix, g, g.GetContainer().GetID())
+		}
+	}
+
+	log.Info("%s total CPU granted: %dm (%d exclusive + %dm shared), total memory granted: %s",
```

Comment:
given that chackAllocations() called quite often, would it be this log.Info() making it too chatty?

Reply:
With policy-debugging off, there would be one such dump
- after startup
- after initial sync
- after reconfiguration
- after allocation requests
- after a release requests
- after resource update requests
- and one after an internal call to reallocateResources()

This last one is called when the configuration changes, so that clearly might be one too much. I can remove that. The others normally change the overall state of allocations, so it might be a good idea to calculate and dump the final state after those.

Reply:
Ok, I updated it so that for `reallocateResources()` we neither dump the pools nor the total allocations by grants, since we later do both as a post-config action.